### PR TITLE
Fix ServerAsyncAuthenticateTest.

### PR DESCRIPTION
**Pull request by baulig**
I don't quite understand how this test could possibly work before without throwing a certificate validation exception in `AuthenticateAsClientAsync()`.

It was calling `new SslStream(clientConnection.GetStream())` to create the client stream (without passing any certificate validation callback), then called `AuthenticateAsClientAsync(serverName,null,clientSslProtocols,false)` on the returned object.  This should fail with a certificate validation exception.

Added a validation callback to the client stream that's accepting any server certificate.

### CODE REVIEW

File: src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs [hunk 1]
```diff
@@ -179,6 +182,16 @@ private static IEnumerable<object[]> ProtocolMismatchData()
             return true;  // allow everything
         }
 
+        // The following method is invoked by the RemoteCertificateValidationDelegate.
+        private bool AllowAnyServerCertificate(
+              object sender,
+              X509Certificate certificate,
+              X509Chain chain,
+              SslPolicyErrors sslPolicyErrors)
+        {
```
Comment (stephentoub) [inline src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs]: I don't understand this change.  This effectively moved the assert that was for the server side to now be on the client side.  I would revert this change and just use `delegate { return true; }` when constructing the client SslStream.

**Comment by stephentoub** [comment]
> I don't quite understand how this test could possibly work before

It would be good to understand that / get to the bottom of it before making changes to the test.  After all, if it was passing before and it shouldn't have, what's to say that the changes are correct, either?

**Comment by baulig** [comment]
The issue was that the client side had no certificate validation callback - so `AuthenticateAsClientAsync()` could not possibly succeed.

File: src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs (hunk 1 above)
Comment (baulig) [reply src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs]: It does not actually move the assert, only the names are flipped.

However, it might really be cleaner to use the delegate that you suggested on the client side and rename the server into `AllowEmptyClientCertificate` - because what's really happening is that the server side's validation function is called with a `null` client certificate.  How does that sound?

**Comment by stephentoub** [comment]
> so AuthenticateAsClientAsync could not possibly succeed

But the test was passing?

File: src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs]: > It does not actually move the assert, only the names are flipped.

Ah, because you not only swapped the names, you also swapped the call sites.  Ok.

> However, it might really be cleaner to use the delegate that you suggested on the client side

Sure, sounds good.

**Comment by baulig** [comment]
I have investigated why the test was previously passing (but failing with Mono) and it turns out that this was actually due to an implementation detail:

In Mono, we perform certificate validation _during_ the handshake and terminate the connection on failure.  This will cause the server-side handshake to fail as well.

I added a comment to the test explaining this.
